[Stable Diffusion Guide] 101 Stable Diffusion Guide directly into the docs (#1927)
huggingface/diffusers: docs/source/en/conceptual/stable_diffusion.mdx

## Intro                                                                                                                                                                                                                                                                                                                                                                                                                                                                                

Stable Diffusion is a [Latent Diffusion model](https://github.com/CompVis/latent-diffusion) developed by researchers from the Machine Vision and Learning group at LMU Munich, *a.k.a* CompVis.                                                                                                                                                                                                                                                                                             

> Model checkpoints were publicly released at the end of August by a collaboration of Stability AI, CompVis, and Runway with support from EleutherAI and LAION. For more information, you can check out [the official blog post](https://stability.ai/blog/stable-diffusion-public-release).                                                                                                                                                                                                  

Reviewer: ```suggestion
Model checkpoints were publicly released at the end of August 2022 by a collaboration of Stability AI, CompVis, and Runway with support from EleutherAI and LAION. For more information, you can check out [the official blog post](https://stability.ai/blog/stable-diffusion-public-release).                                                                                                                                                                                                  
```

Since its public release the community has done an incredible job at working together to make the stable diffusion checkpoints **faster**, **more memory efficient**, and **more performant**.                                                                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                              
🧨 Diffusers offers a simple API to run stable diffusion with all memory, computing, and quality improvements.                                                                                                                                   
                                                                                                                                                                                                                                              

> This notebook walks you through the improvements one-by-one and you can best leverage [`StableDiffusionPipeline`] for **inference**.                                                                                                          

Reviewer: ```suggestion
This notebook walks you through the improvements one-by-one so you can best leverage [`StableDiffusionPipeline`] for **inference**.                                                                                                          
```

When running *Stable Diffusion* in inference, we usually want to generate a certain type, or style of image and then improve upon it. Improving upon a previously generated image means running inference over and over again with a different prompt and potentially a different seed until we are happy with our generation.                                                                                                                                                                 

So to begin with, it is most important to speed up stable diffusion as much as possible to generate as many pictures as possible in a given amount of time.                                                                                   

This can be done by both improving the **computational efficiency** (speed) and the **memory efficiency** (GPU RAM).                                                                                                                          

Let's start by looking into computational efficiency first.                                                                                                                                                                               

> Throughout the notebook, we will focus on [runwayml/stable-diffusion-v1-5](https://huggingface.co/runwayml/stable-diffusion-v1-5):                                                                                                           

Reviewer: 👍

image = pipe(prompt, generator=generator).images[0]                                                                                                                                                                                           
image                                                                                                                                                                                                                                         
```                                                                                                                                                                                                                                           

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/stable_diffusion_101/sd_101_1.png)                                                                                                                                  
                                                                                                                                                                                                                                              

> Cool, this now took roughly 30 seconds on a T4 GPU (you might see faster inference if you're allocated GPU is better than a T4).                                                                                                              

Reviewer: ```suggestion
Cool, this now took roughly 30 seconds on a T4 GPU (you might see faster inference if your allocated GPU is better than a T4).                                                                                                              
```

> The default run we did above used full float32 precision and run the default 50 inference steps. The easiest speed-ups come from switching to float16 (or half) precision and simply running fewer inference steps. Let's load the model now instead of float16.                                                                                                                                                                                                                            

Reviewer: ```suggestion
The default run we did above used full float32 precision and ran the default number of inference steps (50). The easiest speed-ups come from switching to float16 (or half) precision and simply running fewer inference steps. Let's load the model now in float16 instead.                                                                                                                                                                                                                            
```

```                                                                                                                                                                                                                                           
![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/stable_diffusion_101/sd_101_2.png)                                                                                                                                  

Cool, this is almost three times as fast for arguably the same image quality.                                                                                                                                                           
                                                                                                                                                                                                                                              
We strongly suggest always running your pipelines in float16 as so far we have very rarely seen degradations in quality because of it.                                                                                                         

> Next, let's see if we need to use 50 inference steps or whether we could use significantly fewer.                                                                                                                                       

Reviewer: ```suggestion
Next, let's see if we need to use 50 inference steps or whether we could use significantly fewer. The number of inference steps is associated with the denoising scheduler we use. Choosing a more efficient scheduler could help us decrease the number of steps.
```

> Let's have a look at all schedulers, the stable diffusion pipeline is compatible with.                                                                                                                                                        

Reviewer: ```suggestion
Let's have a look at all the schedulers the stable diffusion pipeline is compatible with.                                                                                                                                                        
```

     diffusers.schedulers.scheduling_pndm.PNDMScheduler,                                                                                                                                                                                      
     diffusers.schedulers.scheduling_euler_discrete.EulerDiscreteScheduler,                                                                                                                                                                   
     diffusers.schedulers.scheduling_euler_ancestral_discrete.EulerAncestralDiscreteScheduler,                                                                                                                                                
     diffusers.schedulers.scheduling_dpmsolver_multistep.DPMSolverMultistepScheduler,                                                                                                                                                         
     diffusers.schedulers.scheduling_ddpm.DDPMScheduler,                                                                                                                                                                                      

>      diffusers.schedulers.scheduling_ddim.DDIMScheduler]                                                                                                                                                                                      

Reviewer: Maybe add `DEIS` too, should be similar to `DPMSolverMultistepScheduler` :)

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/stable_diffusion_101/sd_101_3.png)                                                                                                                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                              
The image now does look a little different, but it's arguably still of equally high quality. We now cut inference time to just 4 seconds though 😍. 

## Memory Optimization                                                                                                                                                                                                                        

> More memory indirectly also means more speed since we're often trying to maximize how many images we can generate per second, so more images per inference, more images per second usually.                                                                                                                                                                                                                                                                                                 

Reviewer: ```suggestion
Less memory used in generation indirectly implies more speed, since we're often trying to maximize how many images we can generate per second. Usually, the more images per inference run, the more images per second too.                                                                                                                                              
```

> The easiest way to see how much we can increase memory is to simply try it out, and see when we get a *"Out-of-memory (OOM)"* error.                                                                                                          

Reviewer: ```suggestion
The easiest way to see how many images we can generate at once is to simply try it out, and see when we get a *"Out-of-memory (OOM)"* error.                                                                                                          
```

> One can run batched inference by simply passing a list of prompts and generators. Let's define a quick function that generates a batch for us.                                                                                                

Reviewer: ```suggestion
We can run batched inference by simply passing a list of prompts and generators. Let's define a quick function that generates a batch for us.                                                                                                
```

``` python                                                                                                                                                                                                                                    
def get_inputs(batch_size=1):                                                                                                                                                                                                                 

>   generator = [torch.Generator("cuda").manual_seed(i) for i in range(batch_size)]                                                                                                                                                             

Reviewer: Maybe `generators` ? (also in line 170 if we do change it)

  prompts = batch_size * [prompt]                                                                                                                                                                                                             
  num_inference_steps = 20                                                                                                                                                                                                                    

  return {"prompt": prompts, "generator": generator, "num_inference_steps": num_inference_steps}                                                                                                                                              
```                                                                                                                                                                                                                                           

> 

Reviewer: ```suggestion
This function returns a list of prompts and a list of generators, so we can reuse the generator that produced a result we like.

```

``` python                                                                                                                                                                                                                                    
images = pipe(**get_inputs(batch_size=4)).images                                                                                                                                                                                              
image_grid(images)                                                                                                                                                                                                                            
```                                                                                                                                                                                                                                           

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/stable_diffusion_101/sd_101_4.png)                                                                                                                                  

> Going over a batch_size of 4 will error out in this notebook. Also, we can see we only generate slightly more images per second (3.75s/image) compared to 4s/image previously.                                                                                                                                                                                                                                                                                                                   

Reviewer: ```suggestion
Going over a batch_size of 4 will error out in this notebook (assuming we are running it on a T4 GPU). Also, we can see we only generate slightly more images per second (3.75s/image) compared to 4s/image previously.                                                                                                                                                                                                                                                                                                                   
```

> However, the community has found some nice tricks to improve the memory constraints further. After stable diffusion was released, the community found improvements within days and shared the freely over GitHub - open-source at its finest! I believe the original idea came from [this](https://github.com/basujindal/stable-diffusion/pull/117). GitHub thread.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    

Reviewer: ```suggestion
However, the community has found some nice tricks to improve the memory constraints further. After stable diffusion was released, the community found improvements within days and shared them freely over GitHub - open-source at its finest! I believe the original idea came from [this](https://github.com/basujindal/stable-diffusion/pull/117) GitHub thread.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
```

It can easily be enabled by calling `enable_attention_slicing` as is documented [here](https://huggingface.co/docs/diffusers/main/en/api/pipelines/stable_diffusion/text2img#diffusers.StableDiffusionPipeline.enable_attention_slicing).                                                                                                                                                                                                                                                   

``` python                                                                                                                                                                                                                                    
pipe.enable_attention_slicing()                                                                                                                                                                                                               
```                                                                                                                                                                                                                                           

> Great now that attention slicing is enabled, let's try to double the batch size again, going for `batch_size=8`.                                                                                                                              

Reviewer: ```suggestion
Great, now that attention slicing is enabled, let's try to double the batch size again, going for `batch_size=8`.                                                                                                                              
```

image_grid(images, rows=2, cols=4)                                                                                                                                                                                                            
```                                                                                                                                                                                                                                           

![img](https://huggingface.co/datasets/diffusers/docs-images/resolve/main/stable_diffusion_101/sd_101_5.png)                                                                                                                                  

Nice, it works. However, the speed gain is again not very big (it might however be much more significant on other GPUs).                                                                                                                      

We're at roughly 3.5 seconds per image 🔥 which is probably the fastest we can be with a simple T4 without sacrificing quality.                                                                                                               

> Next, let's look into how to improve the quality!                                                                                                                                                                                             

Reviewer: Maybe mention `xformers`, even if we don't use it in the notebook?

Author: I mentioned it in the very end :-)

Overall, we strongly recommend just trying the models out and reading up on advice online (e.g. it has been shown that using negative prompts is very important for 2.0 and 2.1 to get the highest possible quality. See for example [this nice blog post](https://minimaxir.com/2022/11/stable-diffusion-negative-prompt/).                                                                                                                                                                   

Additionally, the community has started fine-tuning many of the above versions on certain styles with some of them having an extremely high quality and gaining a lot of traction.                                                                                                                                                                                                                                                                                                          

> We recommend simply have a look at all [diffusers checkpoints sorted by downloads and trying out the different checkpoints](https://huggingface.co/models?library=diffusers).                                                                                                                                                                                                                                                                                                               

Reviewer: ```suggestion
We recommend having a look at all [diffusers checkpoints sorted by downloads and trying out the different checkpoints](https://huggingface.co/models?library=diffusers).                                                                                                                                                                                                                                                                                                               
```

Seems like the difference is only very minor, but the new generations are arguably a bit *sharper*.                                                                                                                                           

Cool, finally, let's look a bit into prompt engineering.                                                                                                                                                                                      

Our goal was to generate a photo of an old warrior chief. Let's now try to bring a bit more color into the photos and make the look more impressive.                                                                                          

Originally our prompt was "*portrait photo of an old warrior chief*".                                                                                                                                                                          

> To improve the prompt, it often helps to add cues that could have been used online to save high-quality photos, as well as add more details since.                                                                                         

Reviewer: ```suggestion
To improve the prompt, it often helps to add cues that could have been used online to save high-quality photos, as well as add more details.                                                                                         
```
